packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+        if (member.type === AST_NODE_TYPES.TSIndexSignature) {
+          checkIndexSignature(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+
+        if (
+          member.type === AST_NODE_TYPES.TSMethodSignature ||
+          member.type === AST_NODE_TYPES.TSPropertySignature
+        ) {
+          checkStaticMember(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+      }
+    }
+
+    function checkTuple(
+      param: TSESTree.ArrayPattern,
+      typeAnnotation: TSESTree.TSTupleType,
+    ): void {
+      for (const [index, member] of typeAnnotation.elementTypes.entries()) {
+        // bail on rest type `[...string[]]`
+        if (member.type === AST_NODE_TYPES.TSRestType) {
+          return;
+        }
+
+        const property = param.elements.at(index);
+
+        if (property === undefined) {
+          reportOnMember(member, { type: 'key', key: String(index) });
+          continue;
+        }
+
+        if (
+          // bail in case of `[, ...]`
+          property == null ||
+          // bail on a rest element
+          property.type === AST_NODE_TYPES.RestElement
+        ) {
+          return;
+        }
+
+        checkParam(property, member);
+      }
+    }
+
+    /**
+     * Checks if an index signature property is covered by at least one destructure property.
+     */
+    function checkIndexSignature(
+      member: TSESTree.TSIndexSignature,
+      remainingProperties: RemainingProperties,
+      dynamicProperties: DynamicProperties,
+    ) {
+      const indexParameterType = getConstrainedTypeAtLocation(
+        services,
+        // an index signature must have exactly one parameter, having more
+        // is valid syntax, but invalid typescript
+        member.parameters[0],
+      );
+
+      if (
+        indexSignatureHasMatchingKey(
+          indexParameterType,
+          dynamicProperties,
+          remainingProperties,
+        )
+      ) {
+        return;
+      }
+
+      // Template literal index signatures count as used if any remaining property's type is the
+      // same, this may miss some unused template literal index signatures
+      if (
+        tsutils.isTemplateLiteralType(indexParameterType) &&
+        indexSignatureHasMatchingValue(remainingProperties, member)
+      ) {
+        return;
+      }
+
+      reportOnMember(member, {
+        type: 'index signature',
+        key: `[${checker.typeToString(indexParameterType)}]`,
+      });
+    }
+
+    /**
+     * Checks if an non index-signature property is covered by at least one destructure property.
+     */
+    function checkStaticMember(
+      member: TSESTree.TSMethodSignature | TSESTree.TSPropertySignature,
+      remainingProperties: RemainingProperties,
+      dynamicProperties: DynamicProperties,
+    ): void {
+      const memberKey = getStaticMemberAccessValue(member, context);
+
+      if (memberKey === undefined) {
+        return;
```

Comment:
[Testing] Uncovered line, as reported by codecov. Either this is reachable and should have a test covering it, or it's not and either a refactor or type assertion should be used to eliminate the dead code.